Migration guide: tokenization (#42477)
huggingface/transformers: MIGRATION_GUIDE_V5.md

        )
```

Once the tokenizer is defined as above, you can load it with the following: `Llama5Tokenizer()`. Doing this returns you an empty, trainable tokenizer that follows the definition of the authors of `Llama5` (it does not exist yet :wink:).

The above is the main motivation towards refactoring tokenization: we want tokenizers to behave similarly to models: trained or empty, and with exactly what is defined in their class definition.

### Backend Architecture Changes: moving away from the slow/fast tokenizer separation

Up to now, transformers maintained two parallel implementations for many tokenizers:
- "Slow" tokenizers (`tokenization_<model>.py`) - Python-based implementations, often using [SentencePiece](https://github.com/google/sentencepiece) as the backend.
- "Fast" tokenizers (`tokenization_<model>_fast.py`) - Rust-based implementations using the 🤗 [tokenizers](https://github.com/huggingface/tokenizers) library.

In v5, we consolidate to a single tokenizer file per model: `tokenization_<model>.py`. This file will use the most appropriate backend available:

> 1. **TokenizersBackend** (preferred): Rust-based tokenizers from the 🤗 [tokenizers](https://github.com/huggingface/tokenizers) library. In general it provides optimal performance, but it also offers a lot more features that are commonly adopted across the ecosystem, like handling additional tokens, easily update the state of the tokenizer, automatic parallelization, and automatic offsets, etc. 

Reviewer: ```suggestion
1. **TokenizersBackend** (preferred): Rust-based tokenizers from the 🤗 [tokenizers](https://github.com/huggingface/tokenizers) library. In general it provides optimal performance, but it also offers a lot more features that are commonly adopted across the ecosystem:
  - handling additional tokens
  - a full python API for setting and updating 
  - automatic parallelization,
  - automatic offsets
  - customization
  - training
```

> 2. **SentencePieceBackend**: for tokenizers requiring the `sentencepiece` library

Reviewer: ```suggestion
2. **SentencePieceBackend**: for tokenizers requiring the `sentencepiece` library. It inherits from `PythonBackend`. 
```

> 3. **PythonBackend**: pure Python implementations

Reviewer: ```suggestion
3. **PythonBackend**: a Python implementations of the features provided by `tokenizers`. Basically allows adding tokens.
```

- **MT5** → uses T5Tokenizer
- **MVP** → uses BartTokenizer

These modules will eventually be removed altogether.

**Removed T5-specific workarounds**

The internal `_eventually_correct_t5_max_length` method has been removed. T5 tokenizers now handle max length consistently with other models.

> 

Reviewer 2: ^ sorry I'm commenting instead of 'suggestion' bc as a suggestion it breaks the formatting
